Lib/test/test_pyrepl/test_reader.py

```diff
@@ -279,6 +280,113 @@ def test_prompt_length(self):
         self.assertEqual(prompt, "\033[0;32m樂>\033[0m> ")
         self.assertEqual(l, 5)
 
+    def test_prompt_ps1_raise_exception(self):
+        # Handles exceptions from ps1 prompt
+        class Prompt:
+            def __str__(self): 1/0
+
+        def prepare_reader_keep_prompts(*args, **kwargs):
+            reader = prepare_reader(*args, **kwargs)
+            del reader.get_prompt
+            reader.ps1 = Prompt()
+            reader.ps2 = "... "
+            reader.ps3 = "... "
+            reader.ps4 = ""
+            reader.can_colorize = False
+            reader.paste_mode = False
+            return reader
+
+        events = code_to_events("a=1")
+        reader, _ = handle_events_narrow_console(
+            events,
+            prepare_reader=prepare_reader_keep_prompts,
+        )
```

Comment:
We could use a slightly simpler setup code:
```suggestion
        def prepare_reader(events):
            console = FakeConsole(events)
            config = ReadlineConfig(readline_completer=None)
            reader = ReadlineAlikeReader(console=console, config=config)
            reader.can_colorize = False
            return reader

        events = code_to_events("if some_condition:\nsome_function()")
        reader = prepare_reader(events)
```

Reply:
Unfortunately this doesn't simplify code. I have added more simpler one on my sight. Note, we should use handle_all_events (or handle_events_narrow_console) to process whole input string and get it in the `reader.buffer` to check if '\n' contains or not in buffer.